### PR TITLE
Resolution instead of Bitrate

### PR DESCRIPTION
**Pull request by kaileu**
If you will be ask which Quality you want to use you now can use the Resolution instead of a bitrate

### CODE REVIEW

File: default.py [hunk 1]
```diff
@@ -319,17 +319,21 @@ def ask_bitrate(self, bitrates):
         else:
             return None
 
-    def select_bitrate(self, manifest_bitrates=None):
+    def select_bitrate(self, manifest_bitrates=None, manifest_resolution=None):
         """Returns a bitrate, while honoring the user's /preference/."""
+        manifest_bitrates = manifest_bitrates.keys()
+        print ('Das ist ein Test der Output ist: ', manifest_bitrates)
```
Comment (aqw) [inline default.py]: Can you remove these lines? Or, if you find them useful, please translate them to english and use the logging interface.

---Alex

File: resources/lib/pigskin.py [hunk 2]
```diff
@@ -334,7 +335,7 @@ def get_publishpoint_streams(self, video_id, stream_type=None, game_type=None, u
         request = self.make_request(videoDataPath, 'get')
         akamai_url = xmltodict.parse(request)
         for videoSource in akamai_url['video']['videoSources']['videoSource']:
-            if videoSource['@format']== 'HLS':
+            if videoSource['@format']== 'ChromeCast':
```
Comment (aqw) [inline resources/lib/pigskin.py]: Is this associated with this change?

File: resources/lib/pigskin.py [hunk 3]
```diff
@@ -363,11 +364,15 @@ def get_publishpoint_streams(self, video_id, stream_type=None, game_type=None, u
             if m3u8_obj.is_variant:  # if this m3u8 contains links to other m3u8s
                 for playlist in m3u8_obj.playlists:
                     bitrate = int(playlist.stream_info.bandwidth)
+                    resolution = (str(playlist.stream_info.resolution)).replace(', ','x')
                     streams[bitrate] = m3u8_url[:m3u8_url.rfind('/manifest') + 1] + playlist.uri + '?' + m3u8_url.split('?')[1] + '|' + urllib.urlencode(m3u8_header)
+                    streamsres[bitrate] = resolution
+                    dict = {'url': streams, 'res':streamsres}
```
Comment (aqw) [inline resources/lib/pigskin.py]: I thought "dict" is a reserved word. In any case, this should have a more descriptive name.

File: resources/lib/pigskin.py [hunk 4]
```diff
@@ -363,11 +364,15 @@ def get_publishpoint_streams(self, video_id, stream_type=None, game_type=None, u
...
             else:
                 streams['sole available'] = m3u8_url
 
-        return streams
+        #return streams streamsres
```
Comment (aqw) [inline resources/lib/pigskin.py]: The dict name should be clear enough that it doesn't need to be commented. :-)

**Comment by kaileu** [comment]
I just Display the Codec Information from the M3U8 List.
I am using the Bitrate as in Indice to map These two together at the last function.
I will clean the PR a bit.

File: resources/lib/pigskin.py (hunk 2 above)
Comment (kaileu) [inline resources/lib/pigskin.py]: no but my Git didn't split it...
It might be a rebasing Problem on my side

Comment (aqw) [inline resources/lib/pigskin.py]: When working on multiple fixes, 'git add -p' is a wonderful for choosing exactly which bits you want for a given commit.

---Alex

**Comment by aqw** [comment]
@kaileu What does the information from the m3u8 list look like? I was just confused because you said bitrate -> resolution, but then the variable called 'bitrate' was still used.

I'll wait for your cleaned up version, and then will take a closer look again.

---Alex

**Comment by kaileu** [comment]
I appended a Piece of the Manifest.
```
#EXT-X-STREAM-INF:BANDWIDTH=3671308,RESOLUTION=1280x720,CODECS="avc1.64001f,mp4a.40.2"
QualityLevels(3448310)/Manifest(video,format=m3u8-aapl-v3,audiotrack=aac_UND_2_128,filter=connecttv)
```
You have the BANDWIDTH which is a Number thats changed from stream to stream because akamai is more flexible.
And you have the Resolution if this is always correct i am not quite sure but i think its fine.

But the Resolution isn't a numeric Number which you can sort so i decided to go both ways and use the current way and add a 2nd dictonary with the corresponding Resolution to Display the Resolution instead of the Bitrate because no one helps a Number like 3671308 to Chose a stream from.
1280x720 is way better even if its not unique because akamai messed a bit up with the correct resolutions.

**Comment by aqw** [comment]
@kaileu Thanks for the snippet of the Manifest. I agree: resolution is much more user friendly than bitrate (though a bit less informative), and especially now that their bitrate varies, this makes most sense.

I don't want to nitpick this PR into the ground, but the new variable names you have are still very uninformative. "CombinedDictionary" is basically meaningless. Combined of what?

If you want, I'm fine with making the changes myself, as they're largely cosmetic. But I'm unsure which you'd prefer.

---Alex

**Comment by aqw** [comment]
@kaileu I see you just pushed a bunch more commits. I need to have dinner (and then likely some socializing afterwards...) so I probably won't get to these tonight.

But, before I go, could you please squash all these changes into one commit and force push it? It'll make it easier to review.

Thanks for putting up with all the back-and-forth on this. :-)

---Alex

**Comment by kaileu** [comment]
Of Course if you want you can clean it up.
Yeah i am not a programmer :'D so variables like a and b sounds good for me :-P
(i was still in the Progress of sqashing so yes^^)

File: resources/lib/pigskin.py [hunk 5]
```diff
@@ -363,11 +364,15 @@ def get_publishpoint_streams(self, video_id, stream_type=None, game_type=None, u
             if m3u8_obj.is_variant:  # if this m3u8 contains links to other m3u8s
                 for playlist in m3u8_obj.playlists:
                     bitrate = int(playlist.stream_info.bandwidth)
+                    resolution = (str(playlist.stream_info.resolution)).replace(', ','x')
                     streams[bitrate] = m3u8_url[:m3u8_url.rfind('/manifest') + 1] + playlist.uri + '?' + m3u8_url.split('?')[1] + '|' + urllib.urlencode(m3u8_header)
+                    streamResolutionInformation[bitrate] = resolution
```
Comment (emilsvennesson) [inline resources/lib/pigskin.py]: Please avoid using camel case variable names. We should follow PEP8 as much as possible and this doesn't really fit in well with the rest of the coding style used.

Comment (kaileu) [inline resources/lib/pigskin.py]: As i said before i am not a programmer ^^
So a short summary what PEP8 is would be nice. 
I learned to use camel Case Variables^^

Comment (emilsvennesson) [inline resources/lib/pigskin.py]: It's a style guide on how to write good and readable Python code. It's nothing wrong with using camel case variables: it's just that we should be consistent and not mix naming styles.

**Comment by aqw** [comment]
@kaileu 

So, I reviewed this a bit, and I think the idea of the patch is right, but it should be structured differently.

The manifest bitrates change with every stream, so we can no longer rely on that.

So, my thoughts are:
* the user preference should be changed from preferred bitrate to preferred resolution.
* if the user's preferred resolution has multiple bitrates, use the higher bitrate
* if the user has "ask" selected or their preferred resolution is not in the manifest, provide a "select stream quality" dialog.
  * the dialog's options should be in the format of "resolution (bitrate in Mb/s or Kb/s)" e.g. 1280x720 (6 Mbps)

There's a couple different ways to pass the information around internally, but my intial impulse is a dict of resolutions each containing an array of bitrates. But I imagien that'll become more clear when wading into this.

Also, @emilsvennesson is correct about PEP8. It is considered the standard and "correct" way to code in Python. There's also a few utilities that can check for code for pep8 conformity. I use a command creatively called 'pep8' to help me. :-)

Those are my thoughts on this. Let me know if you have questions, etc. :-)

---Alex

**Comment by kaileu** [comment]
@aqw  the manifest Bitrate changes with every stream that is correct but there are still 11 Quality Classes.
The Bitrate itself varies a bit but still belongs to a certain Quality Class so i don't think ist really neccesary to Change it more as it is already done.

But feel free to make an own implementation of this Issue i don't mind :-).
Most of the more elegant Version of Coding is to complex for me so don't expect a nicer PR ;-)

**Comment by aqw** [comment]
@kaileu Yeah, I saw the Quality Class, but that's unfortunately pretty meaningless to users. :-/

I'll keep this PR open, and will rework this later when I have more time. Unfortunately, all this Game Pass stuff is happening at the same time I have time sensitive projects at work.

---Alex

**Comment by jbarry786** [comment]
@aqw, this may be a dumb question. But on the add-on when you pick a bitrate or you just have "play the highest bitrate" available set... does that mean that the add-on will only try to stream from that particular bitrate?

Is it not possible to have adaptive bitrate streaming to reduce the buffering for some users with lower bandwiths?

With regards to the suggestion presented on this thread, on the games I have tried, these seem to be the resolutions/bitrate for them. Feels like there's no resolution with multiple bitrates.

106          mp4        audio only  106k , mp4a.40.2
267          mp4        256x144     267k , avc1.64000c, mp4a.40.2
369          mp4        384x216     369k , avc1.64000d, mp4a.40.2
523          mp4        384x216     523k , avc1.64000d, mp4a.40.2
727          mp4        512x288     727k , avc1.640015, mp4a.40.2
1085         mp4        640x360    1085k , avc1.64001e, mp4a.40.2
1545         mp4        768x432    1545k , avc1.64001e, mp4a.40.2
2362         mp4        1024x576   2362k , avc1.64001f, mp4a.40.2
3640         mp4        1280x720   3640k , avc1.64001f, mp4a.40.2
4968         mp4        1600x900   4968k , avc1.640028, mp4a.40.2
6246         mp4        1920x1080  6246k , avc1.640028, mp4a.40.2 

------------------------

I suppose you can just display an updated menu, similar to this one:
image


While resolutions are cleaner, bitrates are more informative about the quality of the stream.

P.S: can you please verify if "hiding game length" works? It doesn't seem to work for me. Game length is always displayed for me.

Sorry for the slight offtopic.

**Comment by aqw** [comment]
@jbarry786 Adaptive bitrate streaming is the ideal solution, but it's something we are waiting on Kodi to support. Thankfully, there is active development going on for that feature, so I hope that within the year, we can move to that.

As for the resolutions vs bitrates, the stream list I have examined for some games includes multiple bitrates for a couple resolutions.

Please open an issue for the "hiding game length" bug.

---Alex

**Comment by emilsvennesson** [comment]
In my opinion, the best approach would be to replace the fixed bitrate values with a setting allowing the user to specify the maximum bitrate allowed. Something like this perhaps:
```javascript
def select_bitrate(manifest_bitrates=None):
    """Returns a bitrate while honoring the user's preference."""
    bitrate_setting = int(addon.getSetting('preferred_bitrate'))
    if bitrate_setting == 0:
        preferred_bitrate = 'highest'
    elif bitrate_setting == 1:
        preferred_bitrate = 'limit'
    else:
        preferred_bitrate = 'ask'

    manifest_bitrates.sort(key=int, reverse=True)
    if preferred_bitrate == 'highest':
        return manifest_bitrates[0]
    elif preferred_bitrate == 'limit':
        max_bitrate_allowed = int(addon.getSetting('max_bitrate_allowed'))
        allowed_bitrates = [x for x in manifest_bitrates if max_bitrate_allowed >= int(x)]
        if allowed_bitrates:
            return allowed_bitrates[0]
        else:
            addon_log('No bitrate in stream matched the maximum bitrate allowed.')
            return None
    else:
        return ask_bitrate(manifest_bitrates)
```

I believe that the users that were using the fixed bitrate setting did it because their internet connection couldn't keep up with the highest bitrate. Resolution is not really relevant in regards to that aspect, but I agree that we could display just to be a bit more informative.

**Comment by aqw** [comment]
@emilsvennesson I like the idea of specifying max bitrate. Don't think we can control that in adaptive, but we wouldn't want to anyway.

But a point of correction, all users are using fixed bitrate. It was just those with slower connections and/or monthly bandwidth caps who chose lower bitrates.

---Alex

**Comment by aqw** [comment]
@emilsvennesson  I really like your idea of max bitrate, and think it's probably the better approach. Any chance that you'd be willing to open a PR for that?

---Alex

**Comment by emilsvennesson** [comment]
@aqw I will commit this as soon as possible. :-)

**Comment by aqw** [comment]
@emilsvennesson Great thanks!

**Comment by jm-duke** [comment]
Doing some clean-up and this PR is really old, so I'm closing it.

**Pull request closed without merging**